File: server/lib/monarch/model/relations/inner_join.rb
module Model
  module Relations
    class InnerJoin < Relation
      class << self
        def from_wire_representation(representation, repository)
          left_operand = Relation.from_wire_representation(representation["left_operand"], repository)
          right_operand = Relation.from_wire_representation(representation["right_operand"], repository)
          predicate = Predicates::Predicate.from_wire_representation(representation["predicate"], repository)
          new(left_operand, right_operand, predicate)
        end
      end

      attr_reader :left_operand, :right_operand, :predicate
      def initialize(left_operand, right_operand, predicate)
        @left_operand, @right_operand, @predicate = left_operand, right_operand, predicate
      end

      def column(name)
        left_operand.column(name) || right_operand.column(name)
      end

      def joined_tables
        left_operand.joined_tables + right_operand.joined_tables
      end

      def record_class
        @record_class ||= Class.new(CompositeTuple)
        @record_class.joined_tables = joined_tables
        @record_class
      end

      def build_sql_query(query=SqlQuery.new)
        query.add_condition(predicate)
        left_operand.build_sql_query(query)
        right_operand.build_sql_query(query)
      end

      def build_record_from_database(field_values)
        record_class.new(field_values)
      end
    end
  end
end

File: client/lib/monarch/model/relations/inner_join.js
(function(Monarch) {

Monarch.constructor("Monarch.Model.Relations.InnerJoin", Monarch.Model.Relations.Relation, {

  initialize: function(left_operand, right_operand, predicate) {
    this.left_operand = left_operand;
    this.right_operand = right_operand;
    this.predicate = predicate;
    this.initialize_events_system();
  },

  all_tuples: function() {
    if (this._tuples) return this._tuples;

    var cp = this.cartesian_product();
    window.debug = true;

    return Monarch.Util.select(cp, function(composite_tuple) {
      return this.predicate.evaluate(composite_tuple);
    }.bind(this));
  },

  wire_representation: function() {
    return {
      type: "inner_join",
      left_operand: this.left_operand.wire_representation(),
      right_operand: this.left_operand.wire_representation(),
      predicate: this.predicate.wire_representation()
    };
  },

  // private

  cartesian_product: function() {
    var product = [];
    var self = this;
    Monarch.Util.each(self.left_operand.all_tuples(), function(left_tuple) {
      Monarch.Util.each(self.right_operand.all_tuples(), function(right_tuple) {
        product.push(new Monarch.Model.CompositeTuple(left_tuple, right_tuple));
      });
    })
    return product;
  },

//  evaluate_in_repository: function(repository) {
//    return new Monarch.Model.Relations.Selection(this.operand.evaluate_in_repository(repository), this.predicate);
//  },

//  primary_table: function() {
//    return this.operand.primary_table();
//  },

//  column: function(name) {
//    return this.operand.column(name);
//  },

  // private

  subscribe_to_operands: function() {
    var self = this;
  }
});

})(Monarch);
